book/src/proposals/abi-management.md

```diff
@@ -0,0 +1,434 @@
+# Solana ABI management process
+
+This document proposes Solana ABI management process. The ABI management process
+is an engineering practice and supporting technical framework to avoid
+introducing unintended incompatible ABI changes.
+
+# Problem
+
+The Solana ABI (binary interface to the cluster) is currently only defined 
+implicitly by the implementation, and requires a very careful eye to notice
+breaking changes. This makes it extremely difficult to upgrade the software
+on an existing cluster without rebooting the ledger.
+
+# Requirements and objectives
+
+- Unintended ABI changes can be detected as CI failures mechanically.
+- Newer impl must be able to process oldest data (since genesis) once we go
+  mainnet. (This is in stark contrast with regard to conventional software;
+  A decade-old SSL 3.0 doesn't work anymore, but the genesis block of Bitcoin
+  is still replayable today):
+  - Deserialization stability: Newer implementation (like system programs)
+    processes past object(transactions) exactly in the identical way when
+    replaying the ledger.
+  - Serialization stability: Newer implementation (like AccountsDB) produces
+    newer blobs exactly in the identical way unless its ABI is explicitly and
+    knowingly changed.
+- The objective of this proposal is to protect the ABI while sustaining rather
+  rapid development by opting for mechanical process rather than very long
+  human-driven auditing process. After all, while cluster instability must be
+  avoided, timely delivery of software improvements is desired for all
+  stakeholders of the Solana cluster.
+- Once cryptographically-signed, signed data blob must be identical, so no
+  in-place data format update is possible including inbound and outbound of the
+  online system. Also, considering the sheer volume of transactions we're aiming
+  to handle, retrospective in-place update is undesirable at best.
+
+# Solution
+
+Instead of natural human's eye due-diligence, which should be assumed to fail
+regularly, we need a systematic assurance of not breaking the cluster when
+changing the source code.
+
+For that purpose, we propose a two-fold mechanism of marking every ABI-related
+things in source code (`struct`, `enums`) with the new `#[solana_abi]` attribute
```

Comment:
I'd prefer we not prefix anything with `solana`. No need to brand or namespace it. Just call it what it is.

Reply:
In this case `[solana_abi]` feels pretty accurate.  It's tagging a struct or enum literally is a part of the "Solana ABI"

Reply:
Naming party here! :)

I'll propose new name `frozen_abi`.

Because....:
- Removing `solana_` makes the name just `abi`, which is too generic.
- Previously I added the `solana_` prefix as a namespace because the whole thing are tightly coupled to the Solana via the revision thing (epochs, cluster version, etc). But this doesn't apply anymore.
- Moreover, I could still use `solana_`, but the new proposal is really small in terms of scope and generic enough to be used by other projects. In fact, it even could be a self-standing generic helper crate around the `serde` (I checked if a similar crate does ever exist or not on the Internet). So no justfication for being under `solana_`'s thing.
- One more thing: `frozen` is trendy these days. :mouse: